### PR TITLE
feat: adds a RAML spec generator

### PR DESCRIPTION
**Pull request by achrinza**
Signed-off-by: Rifa Achrinza <25147899+achrinza@users.noreply.github.com>

**This is an early draft PR.**

RAML is an alternate API specification spearheaded by MuleSoft that has its own ecosystems of software and tooling. As a framework, LoopBack 4 should empower developers with their requirements. By supporting RAML, developers will be able to more rapidly integrate LoopBack 4 applications with their RAML software and toolings. This thereby enhances LoopBack 4's accessibility to a wider audience of developers and aids it in standing out as a flexible framework that supports multiple types of APIs.

This packages depends on `webapi-parser` - a package which is maintained by MuleSoft - to convert the OpenAPI 3.0 specification to RAML 1.0.

## Checklist

- [x] DCO (Developer Certificate of Origin) [signed in all commits](https://loopback.io/doc/en/contrib/code-contrib.html)
- [ ] `npm test` passes on your machine
- [ ] New tests added or existing tests modified to cover all changes
- [ ] Code conforms with the [style guide](http://loopback.io/doc/en/contrib/style-guide.html)
- [ ] API Documentation in code was updated
- [ ] Documentation in [/docs/site](../tree/master/docs/site) was updated
- [ ] Affected artifact templates in `packages/cli` were updated
- [ ] Affected example projects in `examples/*` were updated

👉 [Check out how to submit a PR](https://loopback.io/doc/en/lb4/submitting_a_pr.html) 👈

### CODE REVIEW

File: extensions/rest-raml/README.md [hunk 1]
```diff
@@ -0,0 +1,19 @@
+# @loopback/rest-raml
+
```
Comment (bajtos) [inline extensions/rest-raml/README.md:2]: Let's start this extension as an experimental one and add the standard disclaimer:

https://github.com/strongloop/loopback-next/blob/81fab94a16f09da8d6251cef68c33b127861c0d4/extensions/typeorm/README.md#L11-L16

File: extensions/rest-raml/README.md [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+## Basic Use
+
+Configure and load LoopbackRestRamlComponent in the application constructor as
+shown below.
+
+```ts
+import {LoopbackRestRamlComponent} from '@loopback/rest-raml';
+this.component(LoopbackRestRamlComponent);
+```
```
Comment (bajtos) [inline extensions/rest-raml/README.md:19]: Can you please add some instructions now how to obtain the RAML spec/document for application's API? Do we have any plans to add a CLI command for that too?

File: extensions/rest-raml/package.json [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+  "license": "MIT",
+  "files": [
+    "README.md",
+    "dist",
+    "src",
+    "!*/__tests__"
+  ],
+  "peerDependencies": {
+    "@loopback/core": "^2.13.1",
+    "@loopback/rest": "^9.1.1"
```
Comment (bajtos) [inline extensions/rest-raml/package.json:36]: Kudos for using peer dependencies 👍🏻

**Review by bajtos — COMMENTED** [review]
Nice feature!

I don't have strong opinions about the implementation design & details, would you @raymondfeng to chime in?

Personally, I am fine to land an early version and then iterate on the implementation in follow-up pull requests. However, I have few comments to consider even for the initial commit 👇🏻

File: extensions/rest-raml/README.md [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+```ts
+import {RestRamlComponent} from '@loopback/rest-raml';
+this.component(RestRamlComponent);
+```
+
+## Configuration
+
+REST RAML leverages
+[configuration by convention](https://loopback.io/doc/en/lb4/Context.html#configuration-by-convention)
+and is able to generate.
```
Comment (bajtos) [inline extensions/rest-raml/README.md:34]: The sentence seems to end in the middle - what is this extension able to generate?

File: extensions/rest-raml/README.md (hunk 2 above)
Comment (bajtos) [reply extensions/rest-raml/README.md:38]: > Can you please add some instructions on how to obtain the RAML spec/document for application's API? Do we have any plans to add a CLI command for that too?

I think this comment haven't been addressed yet.

File: extensions/rest-raml/src/types.ts [hunk 5]
```diff
@@ -0,0 +1,17 @@
...
+// Node module: @loopback/rest-raml
+// This file is licensed under the MIT License.
+// License text available at https://opensource.org/licenses/MIT
+
+/**
+ * Interface defining the component's options object
+ */
+export interface RestRamlComponentOptions {
+  // Add the definitions here
+  httpPath?: string;
```
Comment (bajtos) [inline extensions/rest-raml/src/types.ts:11]: Please add a tsdoc comment to explain what is the `httpPath` option for? When would the users want to change from the default and what value they should provide?

File: extensions/rest-raml/src/__tests__/regression/component.regression.test.ts [hunk 6]
```diff
@@ -0,0 +1 @@
+// #TODO
```
Comment (bajtos) [inline extensions/rest-raml/src/__tests__/regression/component.regression.test.ts:1]: Please add some code to this test file or else remove it altogether.

File: extensions/rest-raml/package.json [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+  "publishConfig": {
+    "access": "public"
+  },
+  "peerDependencies": {
+    "@loopback/core": "^2.14.1",
+    "@loopback/rest": "^9.1.3"
+  },
+  "dependencies": {
+    "tslib": "^2.0.0",
+    "webapi-parser": "^0.5.0"
```
Comment (bajtos) [inline extensions/rest-raml/package.json:45]: Do you expect to use `webapi-parser` in the production code too? IIUC, this dependency is used in tests only, so maybe it should go to `devDependencies` instead?

**Review by bajtos — COMMENTED** [review]
I quickly skimmed through the new version and posted few comments to consider.

I am not sure what's the status of this work and what kind of feedback you are looking for. Please let me know if there is anything in particular you would like me or others to review.

**Comment by achrinza** [comment]
Thanks for the feedback @bajtos

> I am not sure what's the status of this work and what kind of feedback you are looking for. Please let me know if there is anything in particular you would like me or others to review.

I did not have much time to work on LoopBack 4, hence the lack of progress on the outstanding PRs. Though I should be able to continue in the upcoming days.

File: extensions/rest-raml/package.json (hunk 7 above)
Comment (achrinza) [reply extensions/rest-raml/package.json:45]: `webapi-parser` is the package that converts the OpenAPI 3.0 spec into RAML. This package is a wrapper that triggers the "converter" when the OpenAPI 3.0 spec binding is updated.

**Pull request closed without merging**